### DISCUSSION TITLE
PAGExporter source code?

### DISCUSSION THREAD

**Discussion by alyfreym** [Tencent/libpag#398] [General] [answered]
it is possible to get PAGExporter after effects plugin source code to be able to add new built-in special effects? For example I need support native AE Effect lets call it - "GLITCH".  i need my templates to be dynamic so I have to stop using BMP presynthesis because  it is not support this kind of effect-> Next step I planing to add in to source code support this effect, but I can't find source code of PAGExporter

**Comment by domchen** [comment]
Hi，可以直接中文提问。导出插件不是开源的，请参考FAQ最后一节的详细说明：https://pag.io/docs/faq.html 。简要概括就是，PAG最核心的文件格式编解码模块是开源的，自己从头做一个导出插件并没有技术上的障碍，只有UI工作量。但这部分UI工作量我们没有直接开源的核心原因是：PAG是一个动态演进的格式，我们不希望大家各自添加独立的AE特性支持造成大量的格式分裂，格式新增内容需要有官方统一的管理，否则很容易让PAG变成某个业务自己的私有格式，社区也没法获益。通常情况下如果希望新增一个AE特性，可以给我们提特性需求，在这个文档里记录即可：https://docs.qq.com/sheet/DV2txWlR3Z0lxd1Fp?tab=zh6jef ，只要提出的，我们都会尽快排期去支持，但这里面也还是会有优先级。如果希望参与到PAG官方项目中来加速某个特性的支持，导出插件的部分反而是最小的工作量，大头要做的是在开源的libpag仓库里的codec和rendering目录下，对应增加effect的编码Tag，以及渲染，这里可以参考其他effect的实现。这两部分都可以通过构造数据完成后提交PR合入仓库，只要codec和rendering那部分完善了并且确认是对齐AE效果的，具体AE的导出插件更新会非常快，我们来统一增加并发布文件格式变更就行。

**Comment by odddd-studio** [comment]
妈的 死撒比 插件都不提供 安装又装不上 智障产品

**Comment by domchen** [comment, marked as answer]
PAGExporter插件现在已经完全开源了：https://github.com/Tencent/libpag/tree/main/exporter
